File: Editorial/0101-0200/0112_Path_Sum/C/2_Iterations.c
bool hasPathSum(struct TreeNode *root, int sum) {
    struct TreeNode *node_stack[1000];
    int sum_stack[1000];
    int stack_size = 0;
    if (root == NULL) return false;
    node_stack[stack_size] = root;
    sum_stack[stack_size++] = sum - root->val;
    while (stack_size > 0) {
        struct TreeNode *node = node_stack[--stack_size];
        sum = sum_stack[stack_size];
        if (node->left == NULL && node->right == NULL && sum == 0) return true;
        if (node->right != NULL) {
            node_stack[stack_size] = node->right;
            sum_stack[stack_size++] = sum - node->right->val;
        }
        if (node->left != NULL) {
            node_stack[stack_size] = node->left;
            sum_stack[stack_size++] = sum - node->left->val;
        }
    }
    return false;
}

File: Editorial/0101-0200/0112_Path_Sum/TypeScript/2_Iterations.ts
function hasPathSum(root: TreeNode | null, sum: number): boolean {
    if (!root) return false;
    let nodeStack: TreeNode[] = [];
    let sumStack: number[] = [];
    nodeStack.push(root);
    sumStack.push(sum - root.val);
    while (nodeStack.length > 0) {
        let currentNode = nodeStack.pop();
        let currSum = sumStack.pop();
        if (!currentNode!.left && !currentNode!.right && currSum === 0)
            return true;
        if (currentNode!.right) {
            nodeStack.push(currentNode!.right);
            sumStack.push(currSum! - currentNode!.right.val);
        }
        if (currentNode!.left) {
            nodeStack.push(currentNode!.left);
            sumStack.push(currSum! - currentNode!.left.val);
        }
    }
    return false;
}
